# HG changeset patch
# User Jaroslav Hajek <highegg@gmail.com>
# Date 1237322292 -3600
#      Tue Mar 17 21:38:12 2009 +0100
# Node ID d12f44a5dce38885f8f2a5ab0222f90b1c354b6e
# Parent  c235a59d30a4a072889f973893914de3f1866aff
implement fast octave_value constructor

diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,14 @@
+2009-03-17  Jaroslav Hajek  <highegg@gmail.com>
+
+	* ov.cc (octave_value::octave_value): Move to ov.h
+	* ov.h (octave_value::octave_value): Implement fast inline constructor
+	using a static rep.
+
 2009-03-15  Jaroslav Hajek  <highegg@gmail.com>
 
 	* DLD-FUNCTIONS/cellfun.cc (Fcellslices): New DLD function.
 
 2009-03-13  Jaroslav Hajek  <highegg@gmail.com>
 
 	* ov.h (octave_value::compound_binary_op): Support bool compound ops.
 	* ov.cc (do_binary_op, decompose_binary_op, binary_op_fcn_name):
diff --git a/src/ov.cc b/src/ov.cc
--- a/src/ov.cc
+++ b/src/ov.cc
@@ -469,21 +469,16 @@ octave_value::assign_op_as_string (assig
 
     default:
       retval = "<unknown>";
     }
 
   return retval;
 }
 
-octave_value::octave_value (void)
-  : rep (new octave_base_value ())
-{
-}
-
 octave_value::octave_value (short int i)
   : rep (new octave_scalar (i))
 {
 }
 
 octave_value::octave_value (unsigned short int i)
   : rep (new octave_scalar (i))
 {
diff --git a/src/ov.h b/src/ov.h
--- a/src/ov.h
+++ b/src/ov.h
@@ -151,17 +151,23 @@ public:
 
   static std::string assign_op_as_string (assign_op);
 
   static octave_value empty_conv (const std::string& type,
 				  const octave_value& rhs = octave_value ());
 
   enum magic_colon { magic_colon_t };
 
-  octave_value (void);
+  octave_value (void)
+    {
+      static octave_base_value nil_rep;
+      rep = &nil_rep;
+      rep->count++;
+    }
+
   octave_value (short int i);
   octave_value (unsigned short int i);
   octave_value (int i);
   octave_value (unsigned int i);
   octave_value (long int i);
   octave_value (unsigned long int i);
 
   // FIXME -- these are kluges.  They turn into doubles
